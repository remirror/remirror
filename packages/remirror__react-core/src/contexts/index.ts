export * from './i18n-context';
