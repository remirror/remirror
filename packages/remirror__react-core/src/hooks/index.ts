export * from './use-active';
export * from './use-attrs';
export * from './use-chained-commands';
export * from './use-commands';
export * from './use-current-selection';
export * from './use-doc-changed';
export * from './use-editor-dom-ref';
export * from './use-editor-state';
export * from './use-editor-view';
export * from './use-extension';
export * from './use-extension-event';
export * from './use-force-update';
export * from './use-has-extension';
export * from './use-helpers';
export * from './use-i18n';
export * from './use-manager';
export * from './use-mark-range';
export * from './use-portal-container';
export * from './use-react-framework';
export * from './use-remirror';
export * from './use-remirror-context';
export * from './use-selected-text';
export * from './use-update-reason';
