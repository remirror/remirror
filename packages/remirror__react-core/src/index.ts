export type { I18nProps } from './contexts';
export { I18nProvider } from './contexts';
export type {
  UpdateReason,
  UseExtensionCallback,
  UseI18nReturn,
  UseRemirrorProps,
  UseRemirrorReturn,
} from './hooks';
export {
  RemirrorContext,
  useActive,
  useAttrs,
  useChainedCommands,
  useCommands,
  useCurrentSelection,
  useDocChanged,
  useEditorDomRef,
  useEditorState,
  useEditorView,
  useExtension,
  useExtensionCustomEvent,
  useExtensionEvent,
  useForceUpdate,
  useHasExtension,
  useHelpers,
  useI18n,
  useManager,
  useMarkRange,
  usePortalContainer,
  useRemirror,
  useRemirrorContext,
  useSelectedText,
  useUpdateReason,
} from './hooks';
export type { OnChangeHTMLProps, OnChangeJSONProps } from './on-change';
export { OnChangeHTML, OnChangeJSON } from './on-change';
export { createEditorView } from './prosemirror-view';
export { createReactManager } from './react-helpers';
export type { RemirrorProps } from './react-remirror';
export { EditorComponent, Remirror } from './react-remirror';
export type {
  CreateReactManagerOptions,
  GetRootPropsConfig,
  ReactExtensions,
  ReactFrameworkOutput,
  RefKeyRootProps,
  RefProps,
  UseRemirrorContextType,
} from './react-types';
