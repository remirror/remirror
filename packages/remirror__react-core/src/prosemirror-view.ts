import { DirectEditorProps, EditorView } from '@remirror/pm/view';

/**
 * Creates a new editor view
 *
 * @param place
 * @param props
 */
export function createEditorView(
  place: Node | ((p: HTMLElement) => void) | null,
  props: DirectEditorProps,
): EditorView {
  return new EditorView(place, props);
}
