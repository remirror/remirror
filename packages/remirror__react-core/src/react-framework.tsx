import { Dispatch, ReactNode, Ref, SetStateAction } from 'react';
import {
  AnyExtension,
  ErrorConstant,
  Framework,
  FrameworkOptions,
  FrameworkProps,
  invariant,
  object,
  STATE_OVERRIDE,
  UpdateStateProps,
} from '@remirror/core';
import type { EditorState } from '@remirror/pm/state';
import { EditorView } from '@remirror/pm/view';
import { PlaceholderExtension } from '@remirror/preset-react';

import { composeRefs } from './commonjs-packages/seznam-compose-react-refs';
import type { GetRootPropsConfig, ReactFrameworkOutput, RefKeyRootProps } from './react-types';

export class ReactFramework<Extension extends AnyExtension> extends Framework<
  Extension,
  ReactFrameworkProps<Extension>,
  ReactFrameworkOutput<Extension>
> {
  /**
   * Stores the Prosemirror EditorView dom element.
   */
  #editorRef?: HTMLElement;

  /**
   * Keep track of whether the get root props has been called during the most recent render.
   */
  private readonly rootPropsConfig = {
    called: false,
    count: 0,
  };

  get name() {
    return 'react' as const;
  }

  constructor(props: ReactFrameworkOptions<Extension>) {
    super(props);

    if (this.manager.view) {
      this.manager.view.setProps({
        state: this.manager.view.state,
        dispatchTransaction: this.dispatchTransaction,
        attributes: () => this.getAttributes(),
        editable: () => this.props.editable ?? true,
      });

      return;
    }

    this.manager
      .getExtension(PlaceholderExtension)
      .setOptions({ placeholder: this.props.placeholder ?? '' });
  }

  /**
   * This is called to update props on every render so that values don't become stale.
   */
  update(props: ReactFrameworkOptions<Extension>): this {
    super.update(props);

    return this;
  }

  /**
   * Create the prosemirror editor view.
   */
  protected createView(state: EditorState): EditorView {
    return new EditorView(null, {
      state,
      dispatchTransaction: this.dispatchTransaction,
      attributes: () => this.getAttributes(),
      editable: () => this.props.editable ?? true,
      plugins: [],
    });
  }

  /**
   * The external `getRootProps` that is used to spread props onto a desired
   * holder element for the prosemirror view.
   */
  private readonly getRootProps = <RefKey extends string = 'ref'>(
    options?: GetRootPropsConfig<RefKey>,
  ) => {
    return this.internalGetRootProps(options, null);
  };

  /**
   * Creates the props that should be spread on the root element inside which
   * the prosemirror instance will be rendered.
   *
   * TODO - this is useless - REFACTOR
   */
  private readonly internalGetRootProps = <RefKey extends string = 'ref'>(
    options?: GetRootPropsConfig<RefKey>,
    children?: ReactNode,
  ): RefKeyRootProps<RefKey> => {
    // Ensure that this is the first time `getRootProps` is being called during
    // this commit phase of the .
    // invariant(!this.rootPropsConfig.called, { code: ErrorConstant.REACT_GET_ROOT_PROPS });
    this.rootPropsConfig.called = true;

    const {
      refKey: refKey = 'ref',
      ref,
      ...config
    } = options ?? object<GetRootPropsConfig<RefKey>>();

    return {
      [refKey]: composeRefs(ref as Ref<HTMLElement>, this.onRef),
      key: this.uid,
      ...config,
      children,
    } as any;
  };

  /**
   * Stores the Prosemirror editor dom instance for this component using `refs`.
   */
  private readonly onRef: Ref<HTMLElement> = (element) => {
    if (!element) {
      return;
    }

    this.rootPropsConfig.count += 1;

    invariant(this.rootPropsConfig.count <= 1, {
      code: ErrorConstant.REACT_GET_ROOT_PROPS,
      message: `Called ${this.rootPropsConfig.count} times`,
    });

    this.#editorRef = element;
    this.onRefLoad();
  };

  /**
   * Updates the state either by calling `onChange` when it exists or
   * directly setting the internal state via a `setState` call.
   */
  protected updateState({ state, ...rest }: UpdateStateProps): void {
    const { triggerChange = true, tr, transactions } = rest;

    if (this.props.state) {
      const { onChange } = this.props;

      invariant(onChange, {
        code: ErrorConstant.REACT_CONTROLLED,
        message:
          'You are required to provide the `onChange` handler when creating a controlled editor.',
      });

      invariant(triggerChange, {
        code: ErrorConstant.REACT_CONTROLLED,
        message:
          'Controlled editors do not support `clearContent` or `setContent` where `triggerChange` is `true`. Update the `state` prop instead.',
      });

      if (!this.previousStateOverride) {
        this.previousStateOverride = this.getState();
      }

      this.onChange({ state, tr, transactions });

      return;
    }

    // Check if this is a fresh update directly applied by the developer (without
    // transactions or commands).
    if (!tr && !transactions) {
      state = state.apply(state.tr.setMeta(STATE_OVERRIDE, {}));
    }

    // Update the internal prosemirror state. This happens before we update
    // the component's copy of the state.
    this.view.updateState(state);

    // If `transactions` is an empty array, that means the transaction was cancelled by `filterTransaction`.
    if (triggerChange && transactions?.length !== 0) {
      // Update the `onChange` handler before notifying the manager but only when a change should be triggered.
      this.onChange({ state, tr, transactions });
    }

    this.manager.onStateUpdate({ previousState: this.previousState, state, tr, transactions });
  }

  /**
   * Update the controlled state when the value changes and notify the extension
   * of this update.
   */
  updateControlledState(state: EditorState, previousState?: EditorState): void {
    this.previousStateOverride = previousState;

    // Mark this as a state override so that extensions and plugins know to
    // ignore the transaction.
    state = state.apply(state.tr.setMeta(STATE_OVERRIDE, {}));

    this.view.updateState(state);
    this.manager.onStateUpdate({ previousState: this.previousState, state });

    this.previousStateOverride = undefined;
  }

  /**
   * Adds the prosemirror view to the dom in the position specified via the
   * component props.
   */
  private addProsemirrorViewToDom(element: HTMLElement, viewDom: Element) {
    if (this.props.insertPosition === 'start') {
      element.insertBefore(viewDom, element.firstChild);
    } else {
      element.append(viewDom);
    }
  }

  /**
   * Called once the container dom node (`this.editorRef`) has been initialized
   * after the component mounts.
   *
   * This method handles the cases where the dom is not focused.
   */
  private onRefLoad() {
    invariant(this.#editorRef, {
      code: ErrorConstant.REACT_EDITOR_VIEW,
      message: 'Something went wrong when initializing the text editor. Please check your setup.',
    });

    const { autoFocus } = this.props;
    this.addProsemirrorViewToDom(this.#editorRef, this.view.dom);

    if (autoFocus) {
      this.focus(autoFocus);
    }

    this.onChange();
    this.addFocusListeners();
  }

  /**
   * Called for every update of the props and state.
   */
  onUpdate(): void {
    // Ensure that `children` is still a render prop
    // propIsFunction(this.props.children);

    // Check whether the editable prop has been updated
    if (this.view && this.#editorRef) {
      this.view.setProps({ ...this.view.props, editable: () => this.props.editable ?? true });
    }
  }

  /**
   * Get the framework output.
   */
  get frameworkOutput(): ReactFrameworkOutput<Extension> {
    return {
      ...this.baseOutput,
      getRootProps: this.getRootProps,
      portalContainer: this.manager.store.portalContainer,
    };
  }

  /**
   * Reset the called status of `getRootProps`.
   */
  resetRender(): void {
    // Reset the status of roots props being called
    this.rootPropsConfig.called = false;
    this.rootPropsConfig.count = 0;
  }
}

export interface ReactFrameworkProps<Extension extends AnyExtension>
  extends FrameworkProps<Extension> {
  /**
   * When `onChange` is defined this prop is used to set the next editor
   * state value of the Editor. The value is an instance of the **ProseMirror**
   * [[`EditorState`]].
   *
   * @remarks
   *
   * When this is provided the editor becomes a controlled component. Nothing
   * will be updated unless you explicitly set the value prop to the updated
   * state.
   *
   * Be careful not to set and unset the value as this will trigger an error.
   *
   * When the Editor is set to be controlled there are a number of things to be
   * aware of.
   *
   * - **The last dispatch wins** - Calling multiple dispatches synchronously
   *   during an update is no longer possible since each dispatch needs to be
   *   processed within the `onChange` handler and updated via `setState` call.
   *   Only the most recent call is updated.
   * - **Use chained commands** - These can help resolve the above limitation
   *   for handling multiple updates.
   */
  state?: EditorState | null;

  /**
   * Determine whether the Prosemirror view is inserted at the `start` or `end`
   * of it's container DOM element.
   *
   * @defaultValue 'end'
   */
  insertPosition?: 'start' | 'end';

  /**
   * The placeholder to set for the editor.
   */
  placeholder?: string;
}

/**
 * The options that are passed into the [[`ReactFramework`]] constructor.
 */
export interface ReactFrameworkOptions<Extension extends AnyExtension>
  extends FrameworkOptions<Extension, ReactFrameworkProps<Extension>> {}

export type SetShouldRenderClient = Dispatch<SetStateAction<boolean | undefined>>;
