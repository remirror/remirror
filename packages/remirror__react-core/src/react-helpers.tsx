import { AnyExtension, getLazyArray, RemirrorManager } from '@remirror/core';
import { corePreset } from '@remirror/preset-core';
import { ReactExtension } from '@remirror/preset-react';

import type { CreateReactManagerOptions, ReactExtensions } from './react-types';

/**
 * Create a React [[`RemirrorManager`]] with all the default react presets and
 * extensions.
 */
export function createReactManager<Extension extends AnyExtension>(
  extensions: Extension[] | (() => Extension[]),
  options: CreateReactManagerOptions = {},
): RemirrorManager<ReactExtensions<Extension>> {
  const { core, react, ...settings } = options;

  return RemirrorManager.create(
    () => [...getLazyArray(extensions), new ReactExtension(react), ...corePreset(core)],
    settings,
  );
}
