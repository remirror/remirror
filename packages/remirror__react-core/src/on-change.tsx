import { useCallback } from 'react';
import type { RemirrorJSON } from '@remirror/core';

import { useDocChanged, useHelpers } from './hooks';

export interface OnChangeJSONProps {
  onChange: (json: RemirrorJSON) => void;
}

export const OnChangeJSON = ({ onChange }: OnChangeJSONProps): null => {
  const { getJSON } = useHelpers();

  useDocChanged(
    useCallback(
      ({ state }) => {
        const json = getJSON(state);
        onChange(json);
      },
      [onChange, getJSON],
    ),
  );

  return null;
};

export interface OnChangeHTMLProps {
  onChange: (html: string) => void;
}

export const OnChangeHTML = ({ onChange }: OnChangeHTMLProps): null => {
  const { getHTML } = useHelpers();

  useDocChanged(
    useCallback(
      ({ state }) => {
        const html = getHTML(state);
        onChange(html);
      },
      [onChange, getHTML],
    ),
  );

  return null;
};
