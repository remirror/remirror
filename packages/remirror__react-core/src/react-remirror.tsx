import type { ReactElement, ReactNode } from 'react';
import React from 'react';
import { AnyExtension, isNullOrUndefined, RemirrorManager } from '@remirror/core';
import { RemirrorPortals, usePortals } from '@remirror/extension-react-component';
import { ComponentsTheme } from '@remirror/theme';

import { I18nProps, I18nProvider } from './contexts';
import { RemirrorContext, useReactFramework, useRemirrorContext } from './hooks';
import type { ReactFrameworkProps } from './react-framework';

/**
 * The props for the main `<Remirror />` component.
 */
export interface RemirrorProps<Extension extends AnyExtension = Remirror.Extensions>
  extends Omit<ReactFrameworkProps<Extension>, 'stringHandler' | 'manager'>,
    I18nProps {
  /**
   * This manager composes the extensions provided and provides the
   * functionality used throughout the editor.
   *
   * It is overridden here since there was an issue with type inference when
   * using the manager inherited from `ReactFrameworkProps`.
   */
  manager: RemirrorManager<any>;

  /**
   * The optional children which can be passed into the [`Remirror`].
   */
  children?: ReactNode;

  /**
   * Set this to `start` or `end` to automatically render the editor to the dom.
   *
   * When set to `start` the editor will be added before all other child
   * components. If `end` the editable editor will be added after all child
   * components.
   *
   * When no children are provided the editor will automatically be rendered
   * even without this prop being set.
   *
   * `start` is the preferred value since it helps avoid some of the issues that
   * can arise from `zIndex` issues with floating components rendered within the
   * context.
   *
   * @defaultValue undefined
   */
  autoRender?: boolean | 'start' | 'end';

  /**
   * An array of hooks that can be passed through to the `Remirror` component
   * and will be called in the order provided. Each hook receives no props but
   * will have access to the `RemirrorContext`.
   *
   * If you'd like access to more
   * state, you can wrap the `Remirror` component in a custom provider and
   * attach your state there. It can then be accessed inside the hook via
   * context.
   */
  hooks?: Array<() => void>;
}

/**
 * The default editor placeholder where the prosemirror editor will be rendered.
 */
export const EditorComponent = (): JSX.Element => (
  <div className={ComponentsTheme.EDITOR_WRAPPER} {...useRemirrorContext().getRootProps()} />
);

interface HookComponentProps {
  /**
   * The hook that will be run within the `RemirrorContext`. For access to other
   * contexts, wrap the `<Remirror />` within other contexts and access their
   * values with `useContext`.
   */
  hook: () => void;
}

/**
 * A component which exists to call a prop-less hook.
 */
const HookComponent = (props: HookComponentProps) => {
  props.hook();
  return null;
};

/**
 * [[`Remirror`]] is the component for putting the editor into into it's child
 * component.
 *
 * @remarks
 *
 * The main component for remirror. This acts both as a Provider of the remirror
 * context. All components rendered within `Remirror` have access to the
 * remirror context via `useRemirrorContext`.
 *
 * I can also be rendered as a standalone editor without children. In this case
 * the context can be accessed from outside the editor via
 * `useRemirror().getContext()`.
 */
export function Remirror<Extension extends AnyExtension = Remirror.Extensions>(
  props: RemirrorProps<Extension>,
): ReactElement<RemirrorProps<Extension>> {
  const {
    children,
    autoRender,
    i18nFormat,
    locale,
    supportedLocales,
    hooks = [],
    ...frameworkProps
  } = props;
  const context = useReactFramework(frameworkProps);

  // Subscribe to updates from the [[`PortalContainer`]]
  const portals = usePortals(context.portalContainer);

  // A boolean flag which is true when a default editor should be rendered
  // first. If no children are provided and no configuration is provided for
  // autoRender, the editor will automatically be rendered.
  const autoRenderAtStart =
    autoRender === 'start' || autoRender === true || (!children && isNullOrUndefined(autoRender));

  // Whether to render the editor at the end of the editor.
  const autoRenderAtEnd = autoRender === 'end';

  return (
    <I18nProvider i18nFormat={i18nFormat} locale={locale} supportedLocales={supportedLocales}>
      <RemirrorContext.Provider value={context}>
        <RemirrorPortals portals={portals} />
        {hooks.map((hook, index) => (
          <HookComponent hook={hook} key={index} />
        ))}
        {autoRenderAtStart && <EditorComponent />}
        {children}
        {autoRenderAtEnd && <EditorComponent />}
      </RemirrorContext.Provider>
    </I18nProvider>
  );
}
