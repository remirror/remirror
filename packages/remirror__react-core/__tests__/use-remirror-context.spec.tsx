import { jest } from '@jest/globals';
import { act, renderHook } from '@testing-library/react-hooks';
import { RemirrorTestChain } from 'jest-remirror';
import React, { FC, PropsWithChildren } from 'react';
import { BoldExtension } from 'remirror/extensions';
import { act as renderAct, strictRender } from 'testing/react';

import { createReactManager, Remirror, useRemirrorContext } from '../src';

describe('useRemirrorContext', () => {
  it('returns the provider context', () => {
    const { wrapper } = createTestChain();
    const { result } = renderHook(() => useRemirrorContext({ autoUpdate: true }), { wrapper });

    expect(Object.keys(result.current).sort()).toMatchSnapshot();
    expect(result.current.getState()).toEqual(result.current.getPreviousState());
  });

  it('can listen to updates', () => {
    const { chain, wrapper } = createTestChain();
    const { doc, p } = chain.nodes;
    const { result } = renderHook(() => useRemirrorContext<BoldExtension>({ autoUpdate: true }), {
      wrapper,
    });
    expect(result.current.active.bold()).toBe(false);
    act(() => {
      chain.overwrite(doc(p('Welcome <start>friend<end>')));
    });
    expect(result.current.active.bold()).toBe(false);
    act(() => {
      result.current.commands.toggleBold();
    });
    expect(result.current.active.bold()).toBe(true);
  });

  it('can listen to updates with handler', () => {
    const { chain, wrapper } = createTestChain();
    const { doc, p } = chain.nodes;
    const mock: any = jest.fn();
    renderHook(() => useRemirrorContext<BoldExtension>(mock), { wrapper });
    act(() => {
      chain.overwrite(doc(p('Welcome <start>friend<end>')));
    });
    expect(mock).toHaveBeenCalledTimes(2);
  });

  it('should auto update when in strict mode', () => {
    const mock: any = jest.fn();
    const HookComponent: FC = () => {
      useRemirrorContext({ autoUpdate: true });
      mock();
      return <div />;
    };
    const chain = RemirrorTestChain.create(createReactManager([]));
    strictRender(
      <Remirror manager={chain.manager}>
        <HookComponent />
      </Remirror>,
    );

    for (const char of 'Strict') {
      renderAct(() => {
        chain.insertText(char);
      });
    }

    expect(mock).toHaveBeenCalledTimes(14);
  });
});

/**
 * A helper function for creating the editor setup when testing the editor
 * hooks.
 */
function createTestChain() {
  const chain = RemirrorTestChain.create(createReactManager(() => [new BoldExtension()]));

  const Wrapper: FC<PropsWithChildren<object>> = ({ children }) => (
    <Remirror manager={chain.manager} autoRender>
      {children}
    </Remirror>
  );

  return { chain, wrapper: Wrapper };
}
