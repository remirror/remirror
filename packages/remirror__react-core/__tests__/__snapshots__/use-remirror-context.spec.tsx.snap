// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useRemirrorContext returns the provider context 1`] = `
[
  "active",
  "addHandler",
  "attributes",
  "attrs",
  "blur",
  "chain",
  "clearContent",
  "commands",
  "defaultBlockNode",
  "focus",
  "getExtension",
  "getForcedUpdates",
  "getPluginState",
  "getPreviousState",
  "getRootProps",
  "getState",
  "hasExtension",
  "helpers",
  "manager",
  "markTags",
  "marks",
  "nodeTags",
  "nodes",
  "plainTags",
  "pluginKeys",
  "plugins",
  "portalContainer",
  "schema",
  "setContent",
  "tags",
  "uid",
  "view",
]
`;
